### ISSUE TITLE
actix-web 1.0

### ISSUE THREAD

**Issue by fafhrd91** [actix/actix-web#722]
I am planing to move 0.7 to separate branch and move 1.0 to master in 1 week.

Any thoughts, concerns? 

High level tasks:

- [x] Move 0.7 to separate branch
- [x] Move 1.0 to master
- [x] actix-files tidy up #730
- [x] Migration guide  0.7 -> 1.0
- [x] Upgrade the book
- [x] Upgrade example

**Comment by fafhrd91** [comment]
we should release 0.7.19 at the same time

**Comment by fafhrd91** [comment]
@actix/contributors

**Comment by DoumanAsh** [comment]
> we should release 0.7.19 at the same time

Yes, if you want ping me before preparing 1.0 and I'll publish 0.7.19

Also I think we need to consider making considerable migration guide since 0,7 been there for a while

**Comment by krircc** [comment]
Super excited about this.

**Comment by luover** [comment]
@fafhrd91 I compared it with 0.7.x, There does't multipart、pipeline support.
Another actix support not in this list, Is it  in roadmap?
also The 1.0 is super cool, nice work, thank you

**Comment by fafhrd91** [comment]
@lcowell what do you mean by pipeline? multipart and client are a bit later

**Comment by luover** [comment]
@fafhrd91 I does't look the source code(I will, It is time now), I mean [pipeline](https://github.com/actix/actix-web/blob/master/src/pipeline.rs)

**Comment by fafhrd91** [comment]
pipeline is internal raw http implementation. it is in actix-http now

**Comment by luover** [comment]
thanks for your explanation. I will look into.

**Comment by driftingaway86** [comment]
Excited!

**Comment by cetra3** [comment]
A crazy idea: have you considered renaming this to something else since the new version isn't based upon the `actix` framework?  Like `niko-web` or something 😅.  It would help greatly with ensuring existing blogs/documentation/examples for `0.7` could still be discovered.

Also I am assuming that some features are not implemented yet for the new stack, and some features are not going to be reimplemented.  It would be great to include this distinction in the migration guide.

Are the public sub-dependencies (`actix-net`, etc..) going to be bumped to `1.0` as well?  It would make sense if there are any exposed APIs to ensure lock-step with the main `actix-web` crate.

**Comment by fafhrd91** [comment]
i spend too much time on actix-web, it is not wise to abandon it.
migration doc will contain distinctions. but i expect to bring actors back in several weeks

**Comment by fafhrd91** [comment]
@cetra3 also there is future in 0.7 architecture, it is a dead end.

**Comment by cetra3** [comment]
To clarify:`0.7` will no longer be developed, with the majority of the development taking place in `1.0`?

**Comment by fafhrd91** [comment]
that is correct. but it is already is no longer developed at least for 3 months. i use only actix-net and actix-web 1.0 for my work

**Comment by 0xpr03** [comment]
Is there anything official about actix-web removing actix ? Like why and how that changes things ? I'm a bit surprised.

**Comment by mitsuhiko** [comment]
@0xpr03 the base of actix and actix-web is now actix-rt. It's based on a service abstraction instead of actors now. As of why I think @fafhrd91 can give you the motivation but overall not too much changes unless you sent messages to builtin handlers.

**Comment by 0xpr03** [comment]
Thanks for explaining. It sounded more like actix-web is going to change to something different and I couldn't find much in the PRs and issues. Also I'm currently deciding whether to use 0.7 for the current project or wait till 1.0 has landed.

**Comment by fafhrd91** [comment]
actix-web 0.7 doesnt use actors internally, it just have high level integration with actors. same integration will be available to 1.0 but as external dependency. 

as for reasons to replace actors in actix-web, actors are good for expressing high level components dependencies, but they are too high level for efficient code reuse. actors also imply overhead. actix-net's services on other hand are much lower level, they have very simple interface and provide efficient code re-use design.

**Comment by godofdream** [comment]
How will actix-web 1.0 behave concerning speed. E.g. how will it differ in the techempower benchmarks?

**Comment by fafhrd91** [comment]
I will update techempower next week. I expect 1.0 is slightly faster than 0.7, also it is easy to remove whole app routing layer and make system even faster

**Comment by DoumanAsh** [comment]
If anyone has request for actix static files please put your feature requests to #730 

I'll try to make everything before 1.0 release

**Comment by fafhrd91** [comment]
actix-web 1.0.0-alpha.1 is released

**Comment by fafhrd91** [comment]
@DoumanAsh let's release 0.7.19

**Comment by DoumanAsh** [comment]
@fafhrd91 Sorry just woke up recently, but I guess you already release it, thx 😄

**Comment by fafhrd91** [comment]
alpha 2 is released

**Comment by cetra3** [comment]
Just weighing in on my experience with starting to port one of my apps to 1.0:

I don't know if there's a `Handler` equivalent, but I make heavy use of this trait. All my structs have some local state that doesn't need to be shared with the rest of the app. Not sure what the best migration path is here.

There is no way given a `HttpRequest` struct to get the raw payload or multipart out. I use this pattern a lot. I understand that there is a push towards using extractors for this sort of thing.

The challenge I have been having with extractors is that they are not compile time checked, you can submit any function and it only becomes apparent at runtime. Because I'm a pretty crappy developer, I managed to break it quite a lot! When I first tested this I had the wrong type in my `web::Data` and there wasn't any logs that indicated what I did wrong. @fafhrd91 did help me out here with a change to master, but it still took a lot longer than if the compiler let me know there was a type error

I think extractors are great, and do save a lot of boilerplate but it would also be nice to have some compile time guarantees to save me from myself 😂.

So to summarise my challenges with upgrading:

* No `Handler` trait makes it hard to port my state management as it stands, and would mean a major rewrite if I had to go down the function only path. There might be a better way to do this than just functions though
* Extractors are great but it would be awesome if we could access the same things (payload, multipart) with compile time guarantees

**Comment by Dowwie** [comment]
@fafhrd91:

I agree with @cetra3 on compile-time guarantees for extractors.

As for state, it seems that he's done well with localizing state to a Handler and hasn't needed to share it across App or Route levels.  The requirement is for localized state, not the old Handler pattern.  As to how that would be accomplished is for your consideration.

**Comment by fafhrd91** [comment]
regarding compile-time guarantees for exatractors. i am not sure understand what is it, and what is different solution for that. extractors allow to separate extraction/validation process from handling.
it allow to use declarative style for handlers, i.e you just define type that your handler accepts and that is it. opposite solution is to use request information directly, but that is even more runtime (stringy typed).

**Comment by fafhrd91** [comment]
@cetra3 Replacement for `Handler` trait is `Service`, `Service` is better abstraction than handler, it much more fundamental and you can see actix-net, actix-mqtt, actix-amqp are built on top of same abstraction and it is possible to share code, for example actix-utils::timeoyt::Timeout could be used in any of those packages.

**Comment by fafhrd91** [comment]
next release will be `beta.1`

**Comment by dns2utf8** [comment]
Are you planing to add native-tls again in 1.0?

**Comment by fafhrd91** [comment]
`beta.1` is released

@dns2utf8 i need PR. I don’t use native-tls

**Comment by fafhrd91** [comment]
@cetra3 added `Handler` replacement https://github.com/actix/actix-web/blob/master/src/web.rs#L296

**Comment by fafhrd91** [comment]
`beta.3` is released. next release will be `rc`

**Comment by croigian** [comment]
Why was `data_factory` deleted in `beta.3`?
Is there another way to initialize application data that requires an async runtime?

**Comment by fafhrd91** [comment]
@croigian data_factory complicates app initialization process, I couldn’t implement it in simple way. I am planing to add it back in some form

**Comment by jesskfullwood** [comment]
Excited to hear that `1.0` is close. However, in somewhat unfortunate timing, on the language side `std::future::Future` was recently stabilized and `async/await` is apparently make it into `1.37` (towards the end of August 2019). `tokio` already has [experimental support](https://github.com/tokio-rs/tokio/issues/804) for `async/await`, and it looks like the ecosystem is planning to move over quickly once `1.37` is out.

For `actix-web` to move over will require upgrading to `futures 0.3` which will be a breaking change since `futures 0.1` is part of the public API. Does this mean there is a plan to release a `2.0` version with `async/await` compatibility later this year? The churn will be irritating but I don't see an alternative.

**Comment by 0xpr03** [comment]
@jesskfullwood good point.
Although it's said that there is full backwards compatibility to futures 0.1 it's certainly better to use the upcoming stable implementation with all its benefits.
On another term: I'll go and cry myself to sleep after having an MVP based fully on tokio and futures 0.1..

**Comment by zmlzm** [comment]
I think it is not bad 1.0 like now. the stable is first importment. I think tokio maybe don't use futures 0.3.
Does actix-* will always  dependent on tokio ? It's not good idea, Learning is hard when tokio has more and more libraries.  it also own web frameworks(wrap/tower-web), when they get ready，users maybe  Migrate there to reduce complexity.
Also actix-* can base on std Future to build own async runtime. Then no dependent problem forever.
I think the futures library is for some experiment features for one day into STD, if actix-* target production，It should dependent on std Future and then async/await, no futures library.

**Comment by jesskfullwood** [comment]
The compatibility situation is a bit complex. There are ways to convert `futures 0.3` to `futures 0.1` and vice-versa. But the important consideration is - which one does `tokio` actually 'run' internally? the answer is - currently, it runs a `futures 0.1` Future. In the async/await 'preview', it will accept a `futures 0.3` Future but internally [translates](https://github.com/tokio-rs/tokio/blob/6a8934e897f71a36ee07a818e86a7629bf787db4/tokio/src/async_await.rs#L8) it into a `futures 0.1` Future before running.

I assume that this is temporary and that after `async/await` is stable, `tokio` will change to run `futures 0.3` Futures internally, with `futures 0.1` Futures requiring a compatibility shim. At this point `actix-web` will naturally want to upgrade as well, both to take advantage of `async/await` and to avoid the performance hit that the compatibility shim will incur.

**Comment by cetra3** [comment]
I think that it's fine to release with `futures 0.1`.  The async/await story for std lib is definitely not ready yet, lots of bikeshedding on the syntax is taking place, and a lot of other parts of the ecosystem need to catch up.  It's going to be 6+ months until that all gets fleshed out.

There's nothing stopping releasing 1.0 now, and then when it's all ready, release a 2.0.  It's just a number :)

**Comment by fafhrd91** [comment]
`1.0.0-rc` is released. next release is `1.0`

**Comment by SuperHacker-liuan** [comment]
🎉 actix-web-1.0.0 released!

**Comment by driftingaway86** [comment]
The user guide still references 0.7 https://actix.rs/docs/

**Comment by Lewdwig-V** [comment]
You should probably fix the sample code on the Actix Web front page. 

It does not compile on 1.0/rustc 1.35, complaining that module 'server' is private.

**Comment by mys721tx** [comment]
Would the 1.0 documentation be updated to reflect on the deprecation of `actix_web::dev:Handler` and provide migration guide for `Handler` to `actix_web::dev::Service`?

**Comment by 0xpr03** [comment]
> Would the 1.0 documentation

I think this all falls under the point "update Book to 1.0".
A PR is always welcome! See also https://github.com/actix/actix-website/pull/90

**Comment by zmlzm** [comment]
Look this [Update Tokio to use `std::future`](https://github.com/tokio-rs/tokio/pull/1120), Tokio 0.2 will base on std::future NOT  futures 0.3, It Will bring long-term stability，it will be a huge victory。So, what Actix-* 's Plan? @fafhrd91 Is It will base on Tokio or futures0.3 or std::future ?

**Comment by jesskfullwood** [comment]
To be clear, `futures 0.3` uses `std::future` under the hood. `futures 0.3` is just a bunch of extension traits on top of `std::future`, some of which will eventually end up in `std`. The whole Rust ecosystem (including `tokio`) is moving towards using `std::future`, to take advantage of `async/await`.

**Comment by zmlzm** [comment]
So use `std::future` is the best idea, because `futures 0.3` maybe change api in the future. It's Not 1.0.

**Comment by cldershem** [comment]
> Would the 1.0 documentation be updated to reflect on the deprecation of `actix_web::dev:Handler` and provide migration guide for `Handler` to `actix_web::dev::Service`?

@mys721tx This is the next section to work with actix/actix-website#90. Input is obviously welcome, but right now I am just trying to make a first pass at the book. Try to get all of the code working and lots of notes about things that have changed. After that the PR will definitely need some feedback and work from people that are more intimately familiar with the `actix` ecosystem. I am not really an expert, just someone who got tired of clicking on out dated docs when I googled how to do something! :)
